source/extensions/http/header_formatters/preserve_case/preserve_case_formatter.cc

```diff
@@ -0,0 +1,61 @@
+#include "envoy/extensions/http/header_formatters/preserve_case/v3/preserve_case.pb.h"
+#include "envoy/extensions/http/header_formatters/preserve_case/v3/preserve_case.pb.validate.h"
+#include "envoy/http/header_formatter.h"
+#include "envoy/registry/registry.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace Http {
+namespace HeaderFormatters {
+namespace PreserveCase {
+
+class PreserveCaseHeaderFormatter : public Envoy::Http::StatefulHeaderKeyFormatter {
+public:
+  // Envoy::Http::StatefulHeaderKeyFormatter
+  std::string format(absl::string_view key) const override {
+    const auto remembered_key = original_header_keys_.find(key);
+    if (remembered_key != original_header_keys_.end()) {
+      return *remembered_key;
+    } else {
+      return std::string(key);
```

Comment:
It doesn't look necessary, from the perspective of this particular formatter, to make a string copy here. Either the passed-in string_view, or a view of the set entry, would be returnable as a string_view.

I realize this conforms to an interface defined for a different formatter, but the call-site is actually not the same, I think, so maybe we don't need to have the same signature?

Another optimization I had thought of was not storing anything in the set if the incoming header's casing matches either this-style or This-Style, in which case other stateless formatters could be used.

That all may be in the lower weeds compared to all the other stuff going on when proxying a request but I thought I'd mention it.

Reply:
> That all may be in the lower weeds compared to all the other stuff going on when proxying a request but I thought I'd mention it.

Yeah, I had similar thoughts on perf. I'm not sure it's worth it to hyper optimize this, though there is some low hanging fruit. Let me think about and at least add some TODOs.